src/ess/reduce/nexus/generic_types.py

```diff
@@ -0,0 +1,169 @@
+"""Identifier for an arbitrary monitor"""
+Monitor2 = NewType('Monitor2', int)
+"""Identifier for an arbitrary monitor"""
+Monitor3 = NewType('Monitor3', int)
+"""Identifier for an arbitrary monitor"""
+Monitor4 = NewType('Monitor4', int)
+"""Identifier for an arbitrary monitor"""
+Monitor5 = NewType('Monitor5', int)
+"""Identifier for an arbitrary monitor"""
+Incident = NewType('Incident', int)
+"""Incident monitor"""
+Transmission = NewType('Transmission', int)
+"""Transmission monitor"""
+MonitorType = TypeVar(
+    'MonitorType',
+    Monitor1,
+    Monitor2,
+    Monitor3,
+    Monitor4,
+    Monitor5,
+    Incident,
+    Transmission,
+)
+"""TypeVar used for specifying Incident or Transmission monitor type"""
+
+
+class NeXusMonitorName(sciline.Scope[MonitorType, str], str):
+    """Name of a monitor in a NeXus file."""
+
+
+class NeXusDetector(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Full raw data from a NeXus detector."""
+
+
+class NeXusMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataGroup], sc.DataGroup
+):
+    """Full raw data from a NeXus monitor."""
+
+
+class NeXusSample(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus sample."""
+
+
+class NeXusSource(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus source."""
+
+
+class NeXusDetectorEventData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Data array loaded from a NeXus NXevent_data group within an NXdetector."""
+
+
+class NeXusMonitorEventData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Data array loaded from a NeXus NXevent_data group within an NXmonitor."""
+
+
+class SourcePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the neutron source."""
+
+
+class SamplePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the sample."""
+
+
+class DetectorPositionOffset(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Offset for the detector position, added to base position."""
+
+
+class MonitorPositionOffset(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.Variable], sc.Variable
+):
+    """Offset for the monitor position, added to base position."""
+
+
+class CalibratedDetector(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated data from a detector."""
+
+
+class CalibratedMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated data from a monitor."""
+
+
+class DetectorData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated detector merged with neutron event data."""
+
+
+class MonitorData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated monitor merged with neutron event data."""
+
+
+Component = TypeVar('Component', bound=snx.NXobject)
+
+
+@dataclass
+class NeXusFileSpec(Generic[RunType]):
```

Comment:
Many of the types are defined both in `generic_types.py` and in `types.py`. It does not appear to be an issue. Do you have something in mind?

Reply:
I was just thinking about accidental use of one instead of the other...

Reply:
This is actually possible even without the duplication: If you define `MyType[Generic[T]]`, Sciline treats `MyType` and `MyType[MyArg]` is separate. Maybe this could be caught somehow, since it is a subtle cause for bugs.

Reply:
At least mypy will catch misuse of a generic: `error: Missing type parameters for generic type "G"  [type-arg]`

I think this is not just about simple misuse of a type. It is also about explaining the interface. Duplicating names makes the learning curve steeper. If a user (that can be us) finds both a generic and a regular type of the same name, it is unclear which one to use for a given pipeline.

Essentially, here is a situation I can see finding myself in:
1. Get a pipeline from somewhere.
2. Attempt to `compute` something and get `('No provider found for type', ess.reduce.types.Filename)`.
3. Skim this message and read 'missing Filename'.
4. Search for 'Filename', first result is `generic_types.Filename`.
5. Set `pl[generic_types.Filename] = '...'`.
6. Get the same error.
7. Get mad and proclaim my hate for programming in general and Sciline in particular.

The mistake here was to not read the whole qualname in the error message. This is such an easy mistake to make that I am sure it will happen. Especially when under (time) pressure and thinking about the science, not the code. And that is for me, someone who knows the internals. Imagine a scientist encounters this problem.